### ISSUE TITLE
Question - regarding memory management in state data provider callback

### ISSUE THREAD

**Issue by Dedis23** [sysrepo/sysrepo#1474]
Hello,
I have a question regarding memory management inside a state callback.

My task is to retrieve data from the device via REST whenever the client makes a get request to a particular node in the model.

So I have implemented the a shared library that does the REST part.
It has a function called "get_value_via_REST" that allocates a memory and returns the relevant data (as char*).

In my plugin within a data provider callback of the container "unit-state" my code is:

![Screenshot from 2019-04-21 19-40-49](https://user-images.githubusercontent.com/32620962/56473058-b2c5d300-646e-11e9-94fc-395f2262c9f0.png)

After making a simple test with a netconf client, this does seem to do the job.
However I  want to understand this better in terms of memory management.

Basically my question is that:
When the allocated memory can be freed?
Who will free the memory in: *values?

Also, is there a better way to implement this kind of functionality? as you can see this code is an alternate version of the oven.c example so I learned it from that.

*Note: this model is just for testing purpose as all I do here is just a part of POC for my team.

Thanks in advance,
Dedi

**Comment by michalvasko** [comment]
Hi Dedi,
you are providing the values for *sysrepo* so it takes full ownership of them, you should not free the values yourself, ever. You can use `sr_val_set_str_data()` to properly set the string value.

Regards,
Michal

**Comment by Dedis23** [comment]
Hi Michal!
Thanks
